### PR TITLE
refact(builder) added new methods to volume and container builder

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

**What this PR does / why we need it**:
This PR adds functions required for creating cstor volumes.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 1]
```diff
@@ -193,9 +245,26 @@ func WithArguments(args []string) OptionFunc {
 	}
 }
 
-// WithArguments sets the command arguments of the container
-func (b *Builder) WithVolumeMounts(args []corev1.VolumeMount) *Builder {
-	WithVolumeMounts(args)(b.con)
+// WithVolumeMounts sets the command arguments of the container
+func (b *Builder) WithVolumeMounts(volumeMounts []corev1.VolumeMount) *Builder {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/container/v1alpha1/container.go]: Does this merge or result into duplicates

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 2]
```diff
@@ -205,3 +274,103 @@ func WithVolumeMounts(volumeMounts []corev1.VolumeMount) OptionFunc {
...
+		)
+		return b
+	}
+
+	b.con.ImagePullPolicy = policy
+	return b
+}
+
+// WithPrivilegedSecurityContext sets securitycontext of the container
+func (b *Builder) WithPrivilegedSecurityContext(privilaged *bool) *Builder {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/container/v1alpha1/container.go]: Spell check

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 3]
```diff
@@ -205,3 +274,103 @@ func WithVolumeMounts(volumeMounts []corev1.VolumeMount) OptionFunc {
...
+		return b
+	}
+
+	newresources := *resources
+	b.con.Resources = newresources
+	return b
+}
+
+// WithPorts sets ports of the container
+func (b *Builder) WithPorts(ports []corev1.ContainerPort) *Builder {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/container/v1alpha1/container.go]: Does this merge or result into duplicates. UT is needed.

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 4]
```diff
@@ -205,3 +274,103 @@ func WithVolumeMounts(volumeMounts []corev1.VolumeMount) OptionFunc {
...
+		b.errors = append(
+			b.errors,
+			errors.New(
+				"failed to build container object: missing securitycontext",
+			),
+		)
+		return b
+	}
+
+	newprivilaged := *privilaged
```
Comment (akhilerm) [inline pkg/kubernetes/container/v1alpha1/container.go]: Typo. Its `privileged`

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 5]
```diff
@@ -169,7 +183,26 @@ func WithImage(img string) OptionFunc {
...
+
+	if len(cmd) == 0 {
+		b.errors = append(
+			b.errors,
+			errors.New("failed to build container object: missing command"),
+		)
+		return b
+	}
+
+	newcmd := []string{}
```
Comment (vishnuitta) [inline pkg/kubernetes/container/v1alpha1/container.go]: :+1:

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 1 above)
Comment (shubham14bajpai) [reply pkg/kubernetes/container/v1alpha1/container.go]: Renamed it to WithVolumeMountsNew as it resets the field

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 2 above)
Comment (shubham14bajpai) [reply pkg/kubernetes/container/v1alpha1/container.go]: Done

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 3 above)
Comment (shubham14bajpai) [reply pkg/kubernetes/container/v1alpha1/container.go]: Renamed it to WithPortsNew as it resets the field. Added UT for all new methods

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 4 above)
Comment (shubham14bajpai) [reply pkg/kubernetes/container/v1alpha1/container.go]: Done

**Comment by shubham14bajpai** [comment]
> One question on `VolumeMounts` and `Volume`. Should there be a validation somewhere that, the name of Volume and VolumeMounts are matching?

This check is not required as values are passed through deployment and kubernetes checks for it internally.

**Pull request merged**